packages/jupyter-ai/jupyter_ai/actors.py

```diff
@@ -0,0 +1,185 @@
+            )
+        self.reply_queue.put(agent_message)
+
+
+@ray.remote
+class FileSystemActor():
+    """Processes messages prefixed with /fs. This actor will
+    send the message as input to a RetrieverQA chain, that
+    follows the Retrieval and Generation (RAG) tehnique to
+    query the documents from the index, and sends this context
+    to the LLM to generate the final reply.
+    """
+
+    def __init__(self, reply_queue: Queue):
+        self.reply_queue = reply_queue
+        index_actor = ray.get_actor(ACTOR_TYPE.READ.value)
+        handle = index_actor.get_index.remote()
+        vectorstore = ray.get(handle)
+        if not vectorstore:
+            return
+
+        self.chat_history = []
+        self.chat_provider = ConversationalRetrievalChain.from_llm(
+            OpenAI(temperature=0, verbose=True),
+            vectorstore.as_retriever()
+        )
+
+    def process_message(self, message: HumanChatMessage):
+        query = message.body.split(' ', 1)[-1]
+        
+        index_actor = ray.get_actor(ACTOR_TYPE.READ.value)
+        handle = index_actor.get_index.remote()
+        vectorstore = ray.get(handle)
+        # Have to reference the latest index
+        self.chat_provider.retriever = vectorstore.as_retriever()
+        
+        result = self.chat_provider({"question": query, "chat_history": self.chat_history})
+        reply = result['answer']
+        self.chat_history.append((query, reply)) 
+        agent_message = AgentChatMessage(
+            id=uuid4().hex,
+            time=time.time(),
+            body=reply,
+            reply_to=message.id
+        )
+        self.reply_queue.put(agent_message)
+
+@ray.remote
+class DefaultActor():
+    def __init__(self, reply_queue: Queue):
+        # TODO: Should take the provider/model id as strings
+        
+        self.reply_queue = reply_queue
+        if ChatOpenAINewProvider.auth_strategy.name in os.environ:
+            provider = ChatOpenAINewProvider(model_id="gpt-3.5-turbo")
+            
+            # Create a conversation memory
+            memory = ConversationBufferMemory(return_messages=True)
+            prompt_template = ChatPromptTemplate.from_messages([
+                SystemMessagePromptTemplate.from_template("The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know."),
+                MessagesPlaceholder(variable_name="history"),
+                HumanMessagePromptTemplate.from_template("{input}")
+            ])
+            chain = ConversationChain(
+                llm=provider, 
+                prompt=prompt_template,
+                verbose=True, 
+                memory=memory
+            )
+            self.chat_provider = chain
+
+    def process_message(self, message: HumanChatMessage):
+        if self.chat_provider:
```

Comment:
Is there a reason `chat_provider` wouldn't be created and set? If it is not, what should the behavior be - right not it is silent failure, which may not be ideal.

Reply:
I removed the check, but we need a higher level check somewhere to communicate in the chat UI that the provider is not available, for example if user hasn't added their api keys to environment.